File: apps-tools/streaming_manager/src/streaming_api_lib/examples/dac_example_2.py
#!/usr/bin/python3

import streaming
import numpy as np
import wave
import struct

class Callback(streaming.DACCallback):
    counter = 0

    def sendedPack(self,client, ch1_size, ch2_size):
        #print("Data sent CH1",ch1_size,"CH2",ch2_size)
        self.counter = self.counter + 1

    def connected(self,client,host):
        print("DAC client connected",host)

    def disconnected(self,client,host):
        print("DAC client disconnected",host)

    def error(self,client,host,code):
        print("DAC client error", host, "code" , code)
        client.notifyStop()

    def stopped(self,client,host):
        print("DAC server stopped",host)
        client.notifyStop()

    def stoppedFileEnd(self,client,host):
        print("DAC server stopped. The file is finished",host)
        client.notifyStop()

    def stoppedFileBroken(self,client,host):
        print("DAC server stopped. The file is broken",host)
        client.notifyStop()

    def stoppedEmpty(self,client,host):
        print("DAC server stopped. The file is empty",host)
        client.notifyStop()

    def stoppedMemError(self,client,host):
        print("DAC server stopped",host,". Memory error.")
        client.notifyStop()

    def stoppedMemModify(self,client,host):
        print("DAC server stopped",host,". Memory changed")
        client.notifyStop()

    def stoppedMissingFile(self,client,host):
        print("DAC server stopped",host,". Missing file")
        client.notifyStop()

    def configConnected(self,client,host):
        print("Control client connected",host)
        client.notifyStop()

    def configError(self,client,host,code):
        print("Control client error",host, "code" , code)
        client.notifyStop()

    def configErrorTimeout(self,client,host):
        print("Control client error",host, ". Connection timeout")
        client.notifyStop()

def create_stereo_wav(filename, left_channel, right_channel, sample_rate):
    """Create a stereo WAV file using Python's built-in wave module"""
    with wave.open(filename, 'w') as wav_file:
        # Set parameters: 2 channels, 1 byte per sample (8-bit), sample rate
        wav_file.setnchannels(2)  # stereo
        wav_file.setsampwidth(1)  # 1 byte per sample (8-bit)
        wav_file.setframerate(sample_rate)

        # Interleave left and right channels
        stereo_data = np.empty((left_channel.size + right_channel.size,), dtype=np.uint8)
        stereo_data[0::2] = left_channel
        stereo_data[1::2] = right_channel

        # Convert to bytes and write
        wav_file.writeframes(stereo_data.tobytes())

# Creating a streaming client
obj = streaming.DACStreamClient()

# Creating a callback handler. And also remove the owner, since the client itself will delete the handler.
callback = Callback()
obj.setCallbackFunction(callback.__disown__())

# Disable client logs. They are disabled by default.
obj.setVerbose(True)

# Connect to the server. Do not specify the address. If there is only one server in the network, the client will find it itself.
if (obj.connect() == False):
    print("The client did not connect")
    exit(1)

# Get the current decimation setting
dac_rate = obj.getConfig('dac_rate');
print("Current rate",dac_rate)


# Setting up network mode
obj.sendConfig('dac_pass_mode','DAC_NET')

# Setting up a new decimation setting
obj.sendConfig('dac_rate','125000000')

# Setting the memory block size
obj.sendConfig('block_size','262144')

# Setting the size of reserved memory for DAC streaming
obj.sendConfig('adc_size','2621440')

# Generate sine wave data (8-bit)
frequency = 1
sampling_rate = 1024 * 256
t = np.linspace(0., 1., sampling_rate)
amplitude = np.iinfo(np.int8).max
data = (amplitude * np.sin(2. * np.pi * frequency * t)).astype(np.int8)

# Create stereo data (same signal on both channels)
stereo_left = data
stereo_right = data

# Create WAV file using wave module
create_stereo_wav("sin.wav", stereo_left, stereo_right, sampling_rate)

obj.setRepeatInf(True)

if (obj.startStreamingWAV("./sin.wav")):
    print("Streaming is launched")
else:
    print("Error starting streaming")
    exit(1)

# Waiting for the streaming client to complete its work
obj.wait()

print("Send pack", callback.counter)

File: apps-tools/streaming_manager/src/streaming_api_lib/examples/dac_example_2.cpp
#include <cmath>
#include <csignal>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <memory>
#include <vector>
#include "callbacks.h"
#include "config_streaming.h"
#include "dac_streaming.h"

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#pragma pack(push, 1)
struct WAVHeader {
    uint8_t riff[4] = {'R', 'I', 'F', 'F'};
    uint32_t fileSize;
    uint8_t wave[4] = {'W', 'A', 'V', 'E'};
    uint8_t fmt[4] = {'f', 'm', 't', ' '};
    uint32_t fmtSize = 16;
    uint16_t audioFormat = 1;
    uint16_t numChannels;
    uint32_t sampleRate;
    uint32_t byteRate;
    uint16_t blockAlign;
    uint16_t bitsPerSample;
    uint8_t data[4] = {'d', 'a', 't', 'a'};
    uint32_t dataSize;
};
#pragma pack(pop)

std::shared_ptr<DACStreamClient> obj = nullptr;

class DACCallbackHandler : public DACCallback {
   public:
    int counter = 0;

    void sentPack(DACStreamClient* client, uint32_t ch1_size, uint32_t ch2_size) override { counter++; }

    void connected(DACStreamClient* client, std::string host) override { std::cout << "DAC client connected to " << host << std::endl; }

    void error(DACStreamClient* client, std::string host, int code) override {
        std::cerr << "Error (" << code << ") on host: " << host << std::endl;
        client->notifyStop();
    }

    // Implement all other required callback methods...
};

class ConfigCallbackImpl : public ConfigCallback {
   public:
    void sigInt() override {
        if (obj)
            obj->notifyStop();
    }
};

void writeStereoWAV(const std::string& filename, int sampleRate, const std::vector<int8_t>& left, const std::vector<int8_t>& right) {
    if (left.size() != right.size()) {
        throw std::runtime_error("Channel size mismatch");
    }

    std::ofstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open WAV file");
    }

    WAVHeader header;
    header.sampleRate = sampleRate;
    header.numChannels = 2;
    header.bitsPerSample = 8;
    header.byteRate = sampleRate * header.numChannels * header.bitsPerSample / 8;
    header.blockAlign = header.numChannels * header.bitsPerSample / 8;
    header.dataSize = left.size() * header.numChannels;
    header.fileSize = sizeof(WAVHeader) - 8 + header.dataSize;

    file.write(reinterpret_cast<const char*>(&header), sizeof(header));

    // Interleave samples (LRLRLR...)
    for (size_t i = 0; i < left.size(); ++i) {
        file.write(reinterpret_cast<const char*>(&left[i]), 1);
        file.write(reinterpret_cast<const char*>(&right[i]), 1);
    }
}

std::vector<int8_t> generate8bitSine(double freq, int sampleRate, double duration) {
    std::vector<int8_t> samples;
    const int numSamples = sampleRate * duration;
    const double amplitude = 127.0;

    for (int i = 0; i < numSamples; ++i) {
        double t = static_cast<double>(i) / sampleRate;
        samples.push_back(static_cast<int8_t>(amplitude * std::sin(2.0 * M_PI * freq * t)));
    }

    return samples;
}

int main() {
    auto confClient = std::make_shared<ConfigStreamClient>();
    obj = std::make_shared<DACStreamClient>(confClient);

    auto confCallback = std::make_shared<ConfigCallbackImpl>();
    confClient->addCallback(confCallback);

    auto callback = std::make_shared<DACCallbackHandler>();
    obj->setCallback(callback);
    obj->setVerbose(true);
    confClient->setVerbose(true);

    if (!confClient->connect()) {
        std::cerr << "Connection failed" << std::endl;
        return 1;
    }

    // Configure with larger buffers
    confClient->sendConfig("dac_pass_mode", "DAC_NET");
    confClient->sendConfig("dac_rate", "125000000");
    confClient->sendConfig("block_size", "262144");
    confClient->sendConfig("dac_size", "5242880");

    // Generate stereo WAV
    auto left = generate8bitSine(1.0, 262144, 1.0);
    auto right = generate8bitSine(1.0, 262144, 1.0);
    writeStereoWAV("sin.wav", 262144, left, right);

    // Infinite repeat
    obj->setRepeatInf(true);
    auto host = confClient->getHosts().front();
    if (obj->startStreamingWAV(host, "./sin.wav")) {
        std::cout << "Stereo streaming started" << std::endl;
    } else {
        std::cerr << "Failed to start streaming" << std::endl;
        return 1;
    }

    obj->wait();
    std::cout << "Sent packets: " << callback->counter << std::endl;

    return 0;
}
